core/src/context.rs

```diff
@@ -2,21 +2,120 @@ use error::HolochainError;
 use holochain_agent::Agent;
 use logger::Logger;
 use persister::Persister;
-use std::sync::{Arc, Mutex};
+use state::State;
+use std::sync::{Arc, Mutex, RwLock, RwLockReadGuard};
 
-/// Context holds those aspects of the outside world that a Holochain instance needs to operate
+/// Context holds those aspects of a living Holochain instance that components need to operate.
+/// This includes modules that are injected from the outside like logger and persister
+/// but also the state of the instance that gets injected before passing on the context
+/// to inner components/reducers.
 #[derive(Clone)]
 pub struct Context {
     pub agent: Agent,
     pub logger: Arc<Mutex<Logger>>,
     pub persister: Arc<Mutex<Persister>>,
+    state: Option<Arc<RwLock<State>>>,
 }
 
 impl Context {
+    pub fn new(
+        agent: Agent,
+        logger: Arc<Mutex<Logger>>,
+        persister: Arc<Mutex<Persister>>,
+    ) -> Context {
+        Context {
+            agent,
+            logger,
+            persister,
+            state: None,
+        }
+    }
     // helper function to make it easier to call the logger
     pub fn log(&self, msg: &str) -> Result<(), HolochainError> {
         let mut logger = self.logger.lock().or(Err(HolochainError::LoggingError))?;
         logger.log(msg.to_string());
         Ok(())
     }
+
+    pub fn set_state(&mut self, state: Arc<RwLock<State>>) {
+        self.state = Some(state);
+    }
+
+    pub fn state(&self) -> Option<RwLockReadGuard<State>> {
+        match self.state {
+            None => None,
+            Some(ref s) => Some(s.read().unwrap()),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    extern crate holochain_agent;
+    extern crate test_utils;
+    use super::*;
+    use logger::Logger;
+    use persister::SimplePersister;
+    use state::State;
+    use std::sync::{Arc, Mutex};
+
+    #[derive(Clone, Debug)]
+    pub struct TestLogger {
+        pub log: Vec<String>,
+    }
+
+    impl Logger for TestLogger {
+        fn log(&mut self, msg: String) {
+            self.log.push(msg);
+        }
+    }
+
+    /// create a test logger
+    pub fn test_logger() -> Arc<Mutex<TestLogger>> {
+        Arc::new(Mutex::new(TestLogger { log: Vec::new() }))
+    }
```

Comment:
Duplicate code from `instance.rs`.
Just use `use instance::tests::test_logger;` instead, no?

Reply:
Ah, thanks! I did try using the one from test_utils but that didn't work and I didn't know why. Couldn't find the Logger implementation which was there.. But using the one from instance works.